core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnStringJoin.java

```diff
@@ -0,0 +1,139 @@
+import gov.nist.secauto.metaschema.core.metapath.function.FunctionUtils;
+import gov.nist.secauto.metaschema.core.metapath.function.IArgument;
+import gov.nist.secauto.metaschema.core.metapath.function.IFunction;
+import gov.nist.secauto.metaschema.core.metapath.item.IItem;
+import gov.nist.secauto.metaschema.core.metapath.item.ISequence;
+import gov.nist.secauto.metaschema.core.metapath.item.atomic.IAnyAtomicItem;
+import gov.nist.secauto.metaschema.core.metapath.item.atomic.IStringItem;
+import gov.nist.secauto.metaschema.core.util.ObjectUtils;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Implements the XPath 3.1 <a href=
+ * "https://www.w3.org/TR/xpath-functions-31/#func-string-join">fn:string-join</a>
+ * function.
+ */
+public final class FnStringJoin {
+  private static final String NAME = "string-join";
+  @NonNull
+  static final IFunction SIGNATURE_ONE_ARG = IFunction.builder()
+      .name(NAME)
+      .namespace(MetapathConstants.NS_METAPATH_FUNCTIONS)
+      .deterministic()
+      .contextIndependent()
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("arg1")
+          .type(IAnyAtomicItem.type())
+          .oneOrMore()
+          .build())
+      .returnType(IStringItem.type())
+      .returnOne()
+      .functionHandler(FnStringJoin::execute)
+      .build();
+
+  static final IFunction SIGNATURE_TWO_ARG = IFunction.builder()
+      .name(NAME)
+      .namespace(MetapathConstants.NS_METAPATH_FUNCTIONS)
+      .deterministic()
+      .contextIndependent()
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("arg1")
+          .type(IAnyAtomicItem.type())
+          .zeroOrMore()
+          .build())
+      .argument(IArgument.builder()
+          .name("arg2")
+          .type(IStringItem.type())
+          .zeroOrOne()
+          .build())
+      .returnType(IStringItem.type())
+      .returnOne()
+      .functionHandler(FnStringJoin::execute)
+      .build();
+
+  private FnStringJoin() {
+    // disable construction
+  }
+
+  @SuppressWarnings("unused")
+  @NonNull
+  private static ISequence<IStringItem> execute(
+      @NonNull IFunction function,
+      @NonNull List<ISequence<?>> arguments,
+      @NonNull DynamicContext dynamicContext,
+      IItem focus) {
+
+    ISequence<?> arg1 = arguments.get(0);
+    IStringItem arg2 = arguments.size() == 1 ? IStringItem.valueOf("")
+        : FunctionUtils.asTypeOrNull(arguments.get(1).getFirstItem(true));
+    String separator = arg2 == null ? "" : arg2.asString();
+
+    if (arg1.isEmpty()) {
+      return ISequence.of(IStringItem.valueOf(""));
+    }
+
+    return ISequence.of(stringJoin(ObjectUtils.notNull(arg1.stream()
+        .map(arg -> {
+          assert arg != null;
+          return (IAnyAtomicItem) arg;
+        })), separator));
```

Comment:
This should already be handled by the function call code. It handles type converstion. You can use `FunctionUtils.asType` to cast safely when you initialize arg1 as a `ISequence<? extends IAnyAtomicType>`.

Reply:
Adjusted it https://github.com/metaschema-framework/metaschema-java/pull/302/commits/b17c395cfbff13c16e4e26ae22eee22576a77547, let me know what you think.